File: ios/Classes/ZendeskMessaging.swift
import UIKit
import ZendeskSDKMessaging
import ZendeskSDK

public class ZendeskMessaging: NSObject {
    private static var initializeSuccess: String = "initialize_success"
    private static var initializeFailure: String = "initialize_failure"
    private static var loginSuccess: String = "login_success"
    private static var loginFailure: String = "login_failure"
    private static var logoutSuccess: String = "logout_success"
    private static var logoutFailure: String = "logout_failure"
    
    let TAG = "[ZendeskMessaging]"
    
    private var zendeskPlugin: SwiftZendeskMessagingPlugin? = nil
    private var channel: FlutterMethodChannel? = nil

    init(flutterPlugin: SwiftZendeskMessagingPlugin, channel: FlutterMethodChannel) {
        self.zendeskPlugin = flutterPlugin
        self.channel = channel
    }
    
    func initialize(channelKey: String) {
        print("\(self.TAG) - Channel Key - \(channelKey)\n")
        Zendesk.initialize(withChannelKey: channelKey, messagingFactory: DefaultMessagingFactory()) { result in
            if case let .failure(error) = result {
                self.zendeskPlugin?.isInitialized = false
                print("\(self.TAG) - initialize failure - \(error.localizedDescription)\n")
                self.channel?.invokeMethod(ZendeskMessaging.initializeFailure, arguments: ["error": error.localizedDescription])
            } else {
                self.zendeskPlugin?.isInitialized = true
                print("\(self.TAG) - initialize success")
                self.channel?.invokeMethod(ZendeskMessaging.initializeSuccess, arguments: [:])
            }
        }
    }

    func show(rootViewController: UIViewController?) {
        guard let messagingViewController = Zendesk.instance?.messaging?.messagingViewController() else { return }
        guard let rootViewController = rootViewController else { return }
        rootViewController.present(messagingViewController, animated: true, completion: nil)
        print("\(self.TAG) - show")
    }
    
    func loginUser(jwt: String) {
        Zendesk.instance?.loginUser(with: jwt) { result in
            switch result {
            case .success(let user):
                self.channel?.invokeMethod(ZendeskMessaging.loginSuccess, arguments: ["id": user.id, "externalId": user.externalId])
                break;
            case .failure(let error):
                print("\(self.TAG) - login failure - \(error.localizedDescription)\n")
                self.channel?.invokeMethod(ZendeskMessaging.loginFailure, arguments: ["error": nil])
                break;
            }
        }
    }
    
    func logoutUser() {
        Zendesk.instance?.logoutUser { result in
            switch result {
            case .success:
                self.channel?.invokeMethod(ZendeskMessaging.logoutSuccess, arguments: [])
                break;
            case .failure(let error):
                print("\(self.TAG) - logout failure - \(error.localizedDescription)\n")
                self.channel?.invokeMethod(ZendeskMessaging.logoutFailure, arguments: ["error": nil])
                break;
            }
        }
    }
    func getUnreadMessageCount() -> Int {
        let count = Zendesk.instance?.messaging?.getUnreadMessageCount()
        return count ?? 0
    }
    // Add this method to register for push notifications
    func registerForPushNotifications() {
        let notificationCenter = UNUserNotificationCenter.current()
        notificationCenter.delegate = self
        notificationCenter.requestAuthorization(options: [.alert, .sound, .badge]) { allowed, _ in
            guard allowed else { return }

            DispatchQueue.main.async {
                UIApplication.shared.registerForRemoteNotifications()
            }
        }
    }
    
    // Add this method to update the PushNotifications class with the received device token
    func updatePushNotificationToken(deviceToken: Data) {
        PushNotifications.updatePushNotificationToken(deviceToken)
    }
    
    // UNUserNotificationCenterDelegate methods
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                willPresent notification: UNNotification,
                                withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        let userInfo = notification.request.content.userInfo

        // This checks whether a received push notification should be displayed by Messaging
        let shouldBeDisplayed = PushNotifications.shouldBeDisplayed(userInfo)

        switch shouldBeDisplayed {
        case .messagingShouldDisplay:
            if #available(iOS 14.0, *) {
                completionHandler([.banner, .sound, .badge])
            } else {
                completionHandler([.alert, .sound, .badge])
            }
        case .messagingShouldNotDisplay:
            completionHandler([])
        case .notFromMessaging:
            completionHandler([])
        @unknown default: break
        }
    }

    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                didReceive response: UNNotificationResponse,
                                withCompletionHandler completionHandler: @escaping () -> Void) {
        let userInfo = response.notification.request.content.userInfo

        let shouldBeDisplayed = PushNotifications.shouldBeDisplayed(userInfo)

        switch shouldBeDisplayed {
        case .messagingShouldDisplay:
            PushNotifications.handleTap(userInfo) { viewController in
                // Handle displaying the returned viewController in here
            }
        case .messagingShouldNotDisplay:
            break
        case .notFromMessaging:
            break
        @unknown default: break
        }

        completionHandler()
    }
}

File: android/src/main/kotlin/com/chyiiiiiiiiiiiiii/zendesk_messaging/ZendeskMessaging.kt
import android.content.Intent
import com.chyiiiiiiiiiiiiii.zendesk_messaging.ZendeskMessagingPlugin
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import zendesk.android.Zendesk
import zendesk.android.ZendeskResult
import zendesk.android.ZendeskUser
import zendesk.messaging.android.DefaultMessagingFactory
import zendesk.messaging.android.push.PushNotifications

class ZendeskMessaging(private val plugin: ZendeskMessagingPlugin, private val channel: MethodChannel) {
    companion object {
        const val tag = "[ZendeskMessaging]"

        // Method channel callback keys
        const val initializeSuccess: String = "initialize_success"
        const val initializeFailure: String = "initialize_failure"
        const val loginSuccess: String = "login_success"
        const val loginFailure: String = "login_failure"
        const val logoutSuccess: String = "logout_success"
        const val logoutFailure: String = "logout_failure"
        const val updateTokenSuccess: String = "update_token_success"
        const val updateTokenFailure: String = "update_token_failure"
        const val displayNotificationSuccess: String = "display_notification_success"
        const val displayNotificationFailure: String = "display_notification_failure"
    }


    fun initialize(channelKey: String) {
        println("$tag - Channel Key - $channelKey")
        Zendesk.initialize(
            plugin.activity!!,
            channelKey,
            successCallback = { value ->
                plugin.isInitialized = true;
                println("$tag - initialize success - $value")
                channel.invokeMethod(initializeSuccess, null)
            },
            failureCallback = { error ->
                plugin.isInitialized = false;
                println("$tag - initialize failure - $error")
                channel.invokeMethod(initializeFailure, mapOf("error" to error.message))
            },
            messagingFactory = DefaultMessagingFactory()
        )
    }

    fun show() {
        Zendesk.instance.messaging.showMessaging(plugin.activity!!, Intent.FLAG_ACTIVITY_NEW_TASK)
        println("$tag - show")
    }
    fun getUnreadMessageCount(): Int {
        return try {
            Zendesk.instance.messaging.getUnreadMessageCount()
        }catch (error: Throwable){
            0
        }
    }

    fun loginUser(jwt: String) {
        Zendesk.instance.loginUser(
            jwt,
            { value: ZendeskUser? ->
                value?.let {
                    channel.invokeMethod(loginSuccess, mapOf("id" to it.id, "externalId" to it.externalId))
                } ?: run {
                    channel.invokeMethod(loginSuccess, mapOf("id" to null, "externalId" to null))
                }
            },
            { error: Throwable? ->
                println("$tag - Login failure : ${error?.message}")
                println(error)
                channel.invokeMethod(loginFailure, mapOf("error" to error?.message))
            })
    }

    fun logoutUser() {
        GlobalScope.launch (Dispatchers.Main)  {
            try {
                val result = Zendesk.instance.logoutUser()
                if (result is ZendeskResult.Failure) {
                    channel.invokeMethod(logoutFailure, null)
                } else {
                    channel.invokeMethod(logoutSuccess, null)
                }
            } catch (error: Throwable) {
                println("$tag - Logout failure : ${error.message}")
                channel.invokeMethod(logoutFailure, mapOf("error" to error.message))
            }
        }
    }

    fun updatePushNotificationToken(token: String?) {
        try {
            PushNotifications.updatePushNotificationToken(token)
            channel.invokeMethod(updateTokenSuccess, null)
        } catch (error: Throwable) {
            println("$tag - Update token failure : ${error.message}")
            channel.invokeMethod(updateTokenFailure, mapOf("error" to error.message))
        }
    }

    fun handleRemoteMessage(data: Map<String, String>) {
        val responsibility = PushNotifications.shouldBeDisplayed(data)

        when (responsibility) {
            PushNotifications.PushResponsibility.MESSAGING_SHOULD_DISPLAY -> {
                PushNotifications.displayNotification(plugin.activity!!, data)
                channel.invokeMethod(displayNotificationSuccess, null)
            }
            PushNotifications.PushResponsibility.MESSAGING_SHOULD_NOT_DISPLAY -> {
                // This push belongs to Messaging but it should not be displayed to the end user
            }
            PushNotifications.PushResponsibility.NOT_FROM_MESSAGING -> {
                // This push does not belong to Messaging
            }
        }
    }
}


